### DISCUSSION TITLE
/articles/sharepoint-embedded-create-apps/

### DISCUSSION THREAD

**Discussion by giscus** [Voitanos/voitanos.io-comments#221] [Comments]
Build document-centric apps in this tutorial with SharePoint Embedded. Leverage SharePoint’s document libraries & more Microsoft 365 capabilities!
https://www.andrewconnell.com/articles/sharepoint-embedded-create-apps/

**Comment by SimonSimCity** [comment]
Thanks for the article here! It would be good to add a backslash to the commands, where a new-line was added for readability. By this change, users can copy the command and execute it as is.

**Reply by andrewconnell** [reply]
You're welcome! But, when you said... 

> It would be good to add a backslash to the commands

What do you mean by that?

**Reply by SimonSimCity** [reply]
e.g. in https://www.voitanos.io/blog/sharepoint-embedded-create-apps/#create-the-react-spa-scaffolding, change the command 
```sh
npm install @azure/msal-browser
            @fluentui/react-components
            @fluentui/react-icons
            @microsoft/mgt-react
            @microsoft/mgt-element
            @microsoft/mgt-msal2-provider
            -SE
```
to:

```sh
npm install @azure/msal-browser \
            @fluentui/react-components \
            @fluentui/react-icons \
            @microsoft/mgt-react \
            @microsoft/mgt-element \
            @microsoft/mgt-msal2-provider \
            -SE
```

**Reply by andrewconnell** [reply]
Ah... I see what you're referring to. Good catch... fixed.

**Reply by ZzyzxIsAlreadyTaken** [reply]
What does the backslash do?

**Reply by andrewconnell** [reply]
It's just a line continuation rather than using a newline. Only useful for readability...

**Comment by SimonSimCity** [comment]
When going through the chapter `Creating API endpoints that call endpoints on behalf of a user`, it seems that the callback of `authProvider` receives an object, not a string. In your code later, you change the variable `oboGraphToken` to be extracted from the function but with the name `graphTokenRequest` - that fits again, but here in the context, to me its wrong. Hope this is understandable, as it created a great confusion for me, until I realised that you’re actually doing something different in the code than what you claim to do in the explaining section.

**Reply by andrewconnell** [reply]
Are you saying the way I explain it is a little confusing, or the fact I switched the variable names around in the return types confused you... or are you saying something is wrong with the code? I agree with the first part (variable names/explanation) & will fix that, but...

If you're saying the code is wrong, I'm not sure I agree with that as the project works as intended.

**Reply by SimonSimCity** [reply]
The code is not wrong, it's just that you confused me as a reader in that chapter. Combining the second and third code-section of that chapter, it would look like:

```ts
// get user's access token from the request submitted by the React SPA
const [bearer, token] = (req.headers.authorization || '').split(' ');

// set that token to the assertion and the scopes to the permissions we
//    need in the access token to call Microsoft Graph
const graphTokenRequest = {
  oboAssertion: token,
  scopes: [
    Scopes.GRAPH_SITES_READ_ALL,
    Scopes.SPEMBEDDED_FILESTORAGECONTAINER_SELECTED
  ]
};

// obtain the OBO token using the ConfidentialClientApplication object
const ccaOboResponse = await confidentialClient.acquireTokenOnBehalfOf(graphTokenRequest);
const oboGraphToken = ccaOboResponse!.accessToken;

const authProvider = (callback: MSGraph.AuthProviderCallback) => {
  callback(null, graphTokenRequest);
};

const graphClient = MSGraph.Client.init({
  authProvider: authProvider,
  defaultVersion: 'beta'
});
```

Which doesn't work. In this code-example you'd have to replace `callback(null, graphTokenRequest);` by `callback(null, oboGraphToken);` in order to get it work.

In the later section, that becomes clear, but in that explaining section, you confused me. I had to skip it in order to understand the rest. It's just an unlucky choice that those two variables, which are in a different scope and contain different values, have the same name and are directly in consecutive code-sections in the explaining part.

**Reply by andrewconnell** [reply]
OK, good. That's what I thought. The code has been fixed and we're just at the mercy of the caching gods for a few more hours. I appreciate your feedback!

**Comment by SimonSimCity** [comment]
In the first paragraph of the section `Add Listing Containers to the Web API Project`, you ask the user to create a file called `./server/listContainer.ts`. The file should be called `listContainers.ts` - with an “s” at the end of the name - that's at least the name of the file you try to import later.

**Comment by andyvan27** [comment]
Thanks a lot for a great video and detailed article. I have got your code working on my workstation. 

One missing thing from the downloaded code is to add:

authority: Constants.CLIENT_ENTRA_APP_AUTHORITY to Msal2Provider.

to be able to login with single tenant.

However, you have got it in the step above in this article and I missed it, since I only tried to make your code working rather than following the steps (my bad).

Thanks a again!

**Reply by andrewconnell** [reply]
Thanks for the kind comments and letting me know about the code download. I'll check the code and update the download if necessary. Keep an eye out for more SharePoint Embedded articles coming.

I've got a lot more to share, but have been hard at work preparing for the [SPE Dev Summit](https://techcommunity.microsoft.com/t5/marketplace-community/join-us-for-the-sharepoint-embedded-developer-summit/m-p/4058531) that I'm delivering on behalf of Microsoft next week.

**Comment by akashpanjwani** [comment]
Thanks for the article here. I am facing issue in sharing container with other users in same tenant.
I am always getting bad request. Please help me out with Sharing Container API.

```typescript
export const shareContainer = async (req: Request, res: Response) => {

  if (!req.headers.authorization) {
    res.send(401, { message: 'No access token provided.' });
    return;
  }

  const [bearer, token] = (req.headers.authorization || '').split(' ');

  if (!req.body?.displayName) {
    res.send(400, { message: 'Invalid request: must provide a displayName property in the query parameters or request body' });
    return undefined;
  }

  const [graphSuccess, graphTokenRequest] = await getGraphToken(confidentialClient, token);
  if (!graphSuccess) {
    res.send(200, graphTokenRequest);
    return;
  }
  const authProvider = (callback: MSGraph.AuthProviderCallback) => {
    callback(null, graphTokenRequest);
  };

  try {
    const graphClient = MSGraph.Client.init({
      authProvider: authProvider,
      defaultVersion: 'beta'
    });

    const containerRequestData = {
      "roles": ["writer"],
      "grantedToV2": {
        "user": {
          "userPrincipalName": {{"user email address"}}
        }
      }
    };

    const graphResponse = await graphClient.api(`storage/fileStorage/containers/{{containerID}}/permissions`).post(containerRequestData);

    res.send(200, graphResponse);
    return;
  } catch (error: any) {
    res.send(500, { message: `Failed to share container: ${error.message}` });
    return;
  }
}
```

**Reply by andrewconnell** [reply]
That's not surprising because the code you shared includes two places where you need to set real values: `{{"user email address"}}` & `{{containerID}}`.

If that's not your real code, please share what you're using to better understand the issue, and include details on the response from Microsoft Graph when you get the 400 Bad Request error response.

**Reply by akashpanjwani** [reply]
I have already add it but still showing Bad Request.

```typescript
 const containerRequestData = {
      "roles": ["writer"],
      "grantedToV2": {
        "user": {
          "userPrincipalName": "user@develop.onmicrosoft.com"
        }
      }
    };

const graphResponse = await graphClient.api(`storage/fileStorage/containers/b!xPvDQTk3SkOTsuMMnUyDqsNd8RofVMrs1kkhH/permissions`).post(containerRequestData);
```

**Reply by andrewconnell** [reply]
First, please [fence your code when posting the comment](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks)... it makes it much easier to read (I've edited both your posts to do this). In the future, I'd appreciate it (as would others).

Second, the snippet looks good, but what I can't see because you haven't shared it are the full request details being sent to MS Graph including how you're authenticating (i.e.: app-only / app+user aka: delegated), the full error response, if the user you're doing this with has at least the Manage/Owner role, etc.

**Comment by arenk** [comment]
Thanks for putting this all together!

I’ve successfully created a trial and standard container type a few weeks ago. Since recently, the New-SPOContainerType cmdlet fails with the error message “ContainerType not owned by the tenancy” while establishing a billing relationship (seems to be related to Microsoft.Syntex). Did you ever came across this error message?

**Reply by andrewconnell** [reply]
I haven't run into this error... was there more to the error? You mention that it's related to Microsoft.Syntex... I suspect that was in the error the cmdlet returned. 

If so, you need to enable your M365 tenant for Syntex & the PAYG billing. Once you do that, then you'll be able to associate the new CT with an Azure sub (needs to be the same sub the tenant is configured with. For more info, see [Configure Microsoft Syntex for pay-as-you-go billing](https://learn.microsoft.com/microsoft-365/syntex/syntex-azure-billing).

> [!NOTE]
> I haven't tried this yet, but I ran into the same issue in my tenant when I was trying to set this up for the SPE Dev Summit back in March. It wasn't clear, but in talking with MSFT, that's what seemed to be the blocker.

**Reply by arenk** [reply]
Thanks for your response!

The cmdlet just returned the message below, but I could see in the Azure subscription activity log that the event is related to Microsoft.Syntex. For each container type, it seems that a new Microsoft Syntex account is created in the resource group specified in the cmdlet (in fact I can see the Syntex account for my already existing standard container type if I click "show hidden types" in the resource group overview in the Azure portal).

I really thought the the PAYG billing you mentioned would be the missing piece, but unfortunately it didn't help (although I would assume that I need this as well).

I was also able to get in touch with MSFT in the meantime, they said they can reproduce the issue in one of their environments and will get back to me. I will post an update here as soon as there is a solution. It is also a bit strange since it already worked a few weeks ago with the exact same tenant & subscription.

`New-SPOContainerType : Container type creation failed as we failed to establish a billing relationship. Error:
{"error":{"code":"Forbidden","message":"operation_Id: '88db6ff158848fdb6d06ffcc27d8a52a', message:
'{\"error\":{\"code\":\"accessDenied\",\"message\":\"ContainerType not owned by the tenancy\"}}'"}}`

**Reply by andrewconnell** [reply]
Interesting... yes, please do let me know what you find out from Microsoft as it sounds like a bug... welcome to Public Preview :)

**Reply by arenk** [reply]
There was a cache miss when Azure API looks up the newly created container type (I assume to link it to the Microsoft.Syntex account). MSFT will push a fix to prod soon, should be available in all farms in 4-5 days.

**Reply by andrewconnell** [reply]
Thanks for following up... good to hear it's just a bug that's getting squashed!

**Reply by akashpanjwani** [reply]
I am facing issue in sharing container using below code getting always bad request error as it's working fine in Postman.

```ts
try {
    const graphClient = MSGraph.Client.init({
      authProvider: authProvider,
      defaultVersion: 'beta'
    });

    const containerRequestData = {
      roles: [
        "reader"
      ],
      grantedToV2: {
        user: {
          displayName: "Joni Sherman",
          email: "JoniS@eadevelop.onmicrosoft.com",
          id: "4654b65b-5359-4f7b-be1a-cdf5cabc752d",
          userPrincipalName: "jonis@eadevelop.onmicrosoft.com"
        }
      }
    };

    const graphResponse = await graphClient.api(`storage/fileStorage/containers/b!yEct7krw2Eex_EdwoQm3iFPbrALXCoNJuI9KuXD37MgGsob6l0CTS782hM7UeSo_/permissions`).post(containerRequestData);

    res.send(200, graphResponse);
    return;
  } catch (error: any) {
    res.send(500, { message: `Failed to share container: ${error.message}` });
    return;
  }
```

**Reply by andrewconnell** [reply]
@akashpanjwani said:
> always bad request error

Can you share the details of the error? 

Also, look at the browser's NETWORK tab... can you share the details of the request?

**Comment by StevieBleeds** [comment]
Thank you for this fantastic tutorial Andrew, I've had a lot of fun playing about with it using localhost. 

However, I'm trying to deploy the app to my Azure App Service in my tenant using Azure Pipelines. However, I'm coming up against a bit of a roadblock where the /api endpoint doesn't appear to be served and I'm receiving a 404 error.

I can successfully log into my M365 account using the app, but the app does not display my containers nor does it allow me to create a new one either. I've tried using Advanced Tools in the app service to run the node index.js command and load restify, but the 404 error still persists. I've also started to go down the web.config thread, but have had no luck either. 

My apologies, the React + Node.JS combo is fairly new to me, but any pointers at what I should be looking at would be super helpful

Many thanks,

Steve

**Reply by andrewconnell** [reply]
The download is two different projects: one that implements the server-side API portion (in the **./server** folder), and one that implements the React client-side app (mostly in the **./public** & **./src** folders), but they share dependencies.

Notice in the **package.json** file, there are two scripts: `start:frontend` & `start:backend`. In the **./src/common/constants.ts**, you'll notice the two web apps run on different ports (`http://localhost:3000` for the React app, `http://localhost:3001` for the server-side API).

> This sample wasn't intended to be deployed to production, rather it was designed to be a simple setup to run locally. If you want to deploy this to Azure, you'd have to break it up to deploy to two different web apps or modify how it's built.

That's why you aren't getting any containers or creating them... those are privileged operations that are done by the server-side API... sounds like you only spun up the React app so if you look at the dev tools, you'd notice it's failing when it tries to call it's server part.

**Comment by sumit-valecha** [comment]
thanks for the great article, we will need the Entra AD App on consumer tenant also to register Container Type in consuming tenant,correct??

**Reply by andrewconnell** [reply]
No... you'd just ensure the Entra app you use to manage your SPE app is configured to be multi-tenant.
